src/prof_recent.c

```diff
@@ -506,33 +580,65 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count = prof_recent_alloc_dump_prepare_locked(tsd);
```

Comment:
Can these two lines be moved to right before the for loop? That way we don't need the lock/unlock here.

Reply:
The challenge is, we don't want any flushing under the lock (unless if it's absolutely necessary, like the `if (buf_writer_empty(&buf_writer))` case). It's possible to move these statements right before the loop, but we need to handle any overflow from `emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &dump_max);` and `emitter_json_array_kv_begin(&emitter, "recent_alloc");`. If we really want, we can make sure that the buffer size is always longer than the sum of these two parts, but it doesn't seem to me that there's a clean way to do so.

Reply:
I see. Let's keep it as is.